File: B_Array.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 998244353;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);

    int n, k;
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; ++i) {
        cin >> a[i];
    }

    int max_x = (1LL << k); // 2^K
    vector<int> f(max_x, n + 1); // Initialize with a large value (n+1 is out of bounds)

    // Precompute the smallest index for each possible x
    for (int i = 0; i < n; ++i) {
        for (int mask = 0; mask < max_x; ++mask) {
            int value = mask & a[i];
            if (value != a[i]) {
                f[mask] = min(f[mask], i + 1); // Store 1-based index
            }
        }
    }

    // Calculate the sum of F(x)
    long long result = 0;
    for (int x = 0; x < max_x; ++x) {
        if (f[x] <= n) { // Valid index
            result = (result + f[x]) % MOD;
        }
    }

    cout << result << "\n";
    return 0;
}

File: B_Array.py
MOD = 998244353

def solve(N, K, A):
    max_mask = (1 << K)  
    freq = [0] * max_mask

    for a in A:
        freq[a] += 1

    dp = [0] * max_mask
    for mask in range(max_mask):
        dp[mask] = freq[mask]

    for bit in range(K):  
        for mask in range(max_mask):
            if mask & (1 << bit):  
                dp[mask] += dp[mask ^ (1 << bit)]

    F = [0] * max_mask
    for x in range(max_mask):
        for submask in range(x + 1):
            if (submask & x) == submask and dp[submask] > 0:
                F[x] = submask
                break

    result = sum(F) % MOD
    return result

N, K = map(int, input().split())
A = list(map(int, input().split()))

print(solve(N, K, A))